url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+                    getRegName(rm));
+        } else {
+            throw new IllegalArgumentException("Illegal P,W bits " +
+                    String.format("p:%b, w:%b.", p, w));
+        }
+    }
+
+    /**
+     * アドレシングモード 4 - ロード/ストアマルチプル、
+     * 転送開始アドレスを取得します。
+     *
+     * @param pu    P, U ビット
+     * @param rn    レジスタ番号
+     * @param rlist レジスタリスト
+     * @return 転送開始アドレス
+     */
+    public int getAddrMode4StartAddress(int pu, int rn, int rlist) {
+        switch (pu) {
+        case InstructionARM.PU_ADDR4_IA:
+            return getReg(rn);
+        case InstructionARM.PU_ADDR4_IB:
+            return getReg(rn) + 4;
+        case InstructionARM.PU_ADDR4_DA:
+            return getReg(rn) - (Integer.bitCount(rlist) * 4) + 4;
+        case InstructionARM.PU_ADDR4_DB:
+            return getReg(rn) - (Integer.bitCount(rlist) * 4);
+        default:
+            //do nothing
+            break;
+        }
+
+        throw new IllegalArgumentException("Illegal PU field " +
+                pu + ".");
+    }
+
+    /**
+     * アドレシングモード 4 - ロード/ストアマルチプル、
+     * 転送するデータの長さを取得します。
+     *
+     * @param pu    P, U ビット
+     * @param rlist レジスタリスト
+     * @return 転送するデータの長さ
+     */
+    public int getAddrMode4Length(int pu, int rlist) {
+        switch (pu) {
+        case InstructionARM.PU_ADDR4_IA:
+        case InstructionARM.PU_ADDR4_IB:
+            return Integer.bitCount(rlist) * 4;
+        case InstructionARM.PU_ADDR4_DA:
+        case InstructionARM.PU_ADDR4_DB:
+            return -(Integer.bitCount(rlist) * 4);
+        default:
+            //do nothing
+            break;
+        }
+
+        throw new IllegalArgumentException("Illegal PU field " +
+                pu + ".");
+    }
+
+    /**
+     * ステータスレジスタから汎用レジスタへの転送命令。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeMrs(InstructionARM inst, boolean exec) {
+        boolean r = inst.getBit(22);
+        int sbo = inst.getField(16, 4);
+        int rd = inst.getRdField();
+        int dest;
+
+        if (!exec) {
+            printDisasm(inst,
+                    String.format("mrs%s", inst.getCondFieldName()),
+                    String.format("%s, %s",
+                            getRegName(rd), (r) ? "spsr" : "cpsr"));
+            return;
+        }
+
+        if (!inst.satisfiesCond(getCPSR())) {
+            return;
+        }
+
+        if (sbo != 0xf) {
+            System.out.println("Warning: Illegal instruction, " +
+                    String.format("mrs SBO[19:16](0x%01x) != 0xf.", sbo));
+        }
+
+        if (r) {
+            dest = getSPSR().getValue();
+        } else {
+            dest = getCPSR().getValue();
+        }
+
+        setReg(rd, dest);
+    }
+
+    /**
+     * ステータスレジスタへの値の転送命令。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeMsr(InstructionARM inst, boolean exec) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 1111-1122, lines 1133-1158 into separate methods.